doc/designs/thread-api.md

```diff
@@ -0,0 +1,304 @@
+Thread Pool Support
+===================
+
+OpenSSL wishes to support the internal use of threads for purposes of
+concurrency and parallelism in some circumstances. There are various reasons why
+this is desirable:
+
+  - Some algorithms are designed to be run in parallel (Argon2);
+  - Some transports (e.g. QUIC, DTLS) may need to handle timer events
+    independently of application calls to OpenSSL.
+
+To support this end, OpenSSL can manage an internal thread pool. Tasks can be
+scheduled on the internal thread pool.
+
+There are three options given to an application which wants to use the thread
+pool functionality:
+
+  1. Use OpenSSL's default thread pool functionality.
+
+  2. Use OpenSSL's thread pool functionality with a custom thread spawner
+     method.
+
+  3. Provide threads manually to OpenSSL for use as part of its thread pool.
+
+Default Model
+-------------
+
+In the default model, OpenSSL creates and manages threads up to a maximum
+number of threads authorized by the application.
+
+The application enables thread pooling by calling the following function
+during its initialisation:
+
+```c
+/*
+ * Set the maximum number of threads to be used by the thread pool.
+ *
+ * If the argument is 0, thread pooling is disabled. OpenSSL will not create any
+ * threads and existing threads in the thread pool will be torn down.
+ *
+ * Returns 1 on success and 0 on failure. Returns failure if OpenSSL-managed
+ * thread pooling is not supported (for example, if it is not supported on the
+ * current platform, or because OpenSSL is not built with the necessary
+ * support).
+ */
+int OSSL_set_max_threads(uint32_t max_threads);
```

Comment:
Why only uint32_t?
Modern supercomputers are well on the way there already.

Our API stability policy is pretty harsh if we get something wrong :cry:

Reply:
Should we go with uint64_t or size_t?

Reply:
uint64_t IMO.  We're rather unlikely to ever have that many processors.

Reply:
I'd go with size_t as internally you will cast it anyway to it I think.

Reply:
size_t is defined around the number of bytes in the address space.  It seems kind of inappropriate.  Still, I don't care enough either way to argue about it.